plugins/esbuild.ts

```diff
@@ -225,61 +233,94 @@ export function esbuild(userOptions?: Options) {
+            normalizePath(file.path).replace(basePath, ""),
           );
 
-          return outdir === urlWithoutExt;
+          return relativeFilePath === normalizedOutPath;
         });
+        if (!outputFile) {
+          log.error(
+            `[esbuild plugin] Could not match the metafile ${normalizedOutPath} to an output file.`,
+          );
+          continue;
+        }
+
+        // Replace .tsx and .jsx extensions with .js
+        const content = (!options.options.splitting && !options.options.bundle)
+          ? resolveImports(outputFile.text, options.esm)
+          : outputFile.text;
 
         // Get the associated source map
         const map = enableSourceMap
-          ? outputFiles.find((f) => f.path === `${file.path}.map`)
+          ? outputFiles.find((f) => f.path === `${outputFile.path}.map`)
           : undefined;
 
-        // The page is an entry point
-        if (entryPoint) {
-          const outExt = getOutExtension(ext, outExtension);
-          entryPoint.data.url = posix.join(
-            "/",
-            options.options.outdir || ".",
-            replaceExtension(entryPoint.data.url, outExt),
+        // The page is a chunk
+        if (!output.entryPoint) {
+          const page = Page.create({ url: normalizedOutPath });
+          saveAsset(site, page, content, map?.text);
+          allPages.push(page);
+          continue;
+        }
+
+        let outputRelativeEntryPoint = output.entryPoint;
+        if (outputRelativeEntryPoint.startsWith("deno:")) {
+          outputRelativeEntryPoint = outputRelativeEntryPoint.slice(
+            "deno:".length,
+          );
+        }
+        if (outputRelativeEntryPoint.startsWith(prefix)) {
+          outputRelativeEntryPoint = outputRelativeEntryPoint.slice(
+            prefix.length,
+          );
+        }
+        if (outputRelativeEntryPoint.startsWith("file://")) {
+          outputRelativeEntryPoint = fromFileUrl(outputRelativeEntryPoint);
+        }
+        outputRelativeEntryPoint = normalizePath(
+          normalizePath(outputRelativeEntryPoint).replace(basePath, ""),
+        );
+
+        // Search the entry point of this output file
+        const entryPoint = pages.find((page) =>
+          page.sourcePath === outputRelativeEntryPoint
+        );
+        if (!entryPoint) {
+          log.error(
+            `[esbuild plugin] Could not match the entrypoint ${outputRelativeEntryPoint} of metafile ${normalizedOutPath} to an page.`,
           );
-          saveAsset(site, entryPoint, content, map?.text);
           continue;
         }
 
-        // The page is a chunk
-        const page = Page.create({ url });
-        saveAsset(site, page, content, map?.text);
-        allPages.push(page);
+        // The page is an entry point
+        entryPoint.data.url = normalizedOutPath.replaceAll(
+          dirname(entryPoint.sourcePath) + "/",
+          dirname(entryPoint.data.url) + "/",
+        );
```

Comment:
As you commented, this won't work in all cases (like with dynamic imports). I think the proper way to handle this is by using the `outdir` or `outbase` options.

Reply:
Hm, Im not sure how `outdir` or `outbase` would help here. Maybe you could elaborate? From my understanding: We need to invoke esbuild with all entry points at once (for code splitting and rewriting of dynamic imports). The `outdir` and `outbase` are both static string options. Assuming our entry points are in the same subdirectory, we could change the `outbase` to that subdirectory and use the subdirectory path as the `outdir`. This should however not help since esbuild uses relative paths for dynamic imports anyways. What we actually need is a way to _dynamically_ set/overwrite the output path for an esbuild entry point. Or at least a way to overwrite the `[dir]` part of the `entryNames` option. Something like this:

```js
esbuild.build({
  entryNames: (entryFile) => lume.getPage(entryFile.path).data.url
})
```
Howerver esbuild currently does not offer a way to dynamically set the output path.

Reply:
Hmm, okay. I was thinking of the case where all files are in the same directory (for example, `/scripts/`) and the folder has a `_data.yml` file to change the basename (for example to `/app/`). In this case, I guess it could be possible to use root base folder of all entry points (set the outdir to `_site/app/`). But it won't work in all cases.

Let's try your code.

Reply:
Dynamic imports in combination with a `basename` url overwrite generally do work (with your proposal or with the PR). Because esbuild uses relative paths for dynamic imports. With your src example:

- scripts/
  - _data.yml
    - `basename: app`
  - index.ts
    - `import("./big-library.ts")`
  - big-library.ts

Would output:

- app/
  - index.js
    - `import("./big-library.js")`
  - big-library.js

What does not work (neither with your proposal nor the PR) is dynamically importing a script from another folder with a url overwrite:

- scripts/
  - index.ts
    - `import("./dynamic/big-library.ts")`
  - dynamic/
    - _data.yml
      - `basename: other`
    - big-library.ts

Would output:

- app/
  - index.js
    - `import("./dynamic/big-library.js")` // ERR: not found
  - other/
    - big-library.js